Sources/PackageGraph/PackageGraphRoot.swift

```diff
@@ -33,23 +33,33 @@ public struct PackageGraphRootInput {
 /// Represents the inputs to the package graph.
 public struct PackageGraphRoot {
 
-    /// The list of root manifests.
-    public let manifests: [Manifest]
+    /// The root packages.
+    public let packages: [PackageIdentity: (manifest: Manifest, packageReference: PackageReference)]
+
+    /// The root manifests.
+    public var manifests: [PackageIdentity: Manifest] {
+        return self.packages.mapValues { $0.manifest }
+    }
 
     /// The root package references.
-    public let packageRefs: [PackageReference]
+    public var packageReferences: [PackageReference] {
+        return self.packages.values.map { $0.packageReference }
+    }
 
     /// The top level dependencies.
     public let dependencies: [PackageDependencyDescription]
 
     /// Create a package graph root.
-    public init(input: PackageGraphRootInput, manifests: [Manifest], explicitProduct: String? = nil) {
-        // TODO: this does not use the identity resolver which is probably fine since its the root packages
-        self.packageRefs = zip(input.packages, manifests).map { (path, manifest) in
-            let identity = PackageIdentity(url: manifest.packageLocation)
-            return .root(identity: identity, path: path)
-        }
-        self.manifests = manifests
+    public init(input: PackageGraphRootInput, manifests: [Manifest], explicitProduct: String? = nil) throws {
+        self.packages = try input.packages.reduce(into: .init(), { partial, inputPath in
+            let manifestPath = inputPath.basename == Manifest.filename ? inputPath : inputPath.appending(component: Manifest.filename)
+            let packagePath = manifestPath.parentDirectory
+            guard let manifest = (manifests.first{ $0.path == manifestPath}) else {
+                throw InternalError("manifest for \(inputPath) not found")
+            }
+            let identity = PackageIdentity(path: packagePath) // this does not use the identity resolver which is fine since these are the root packages
+            partial[identity] = (manifest, .root(identity: identity, path: packagePath))
+        })
```

Comment:
There were two parts.  The first was this:
```
        guard !diagnostics.hasErrors else {
             throw Diagnostics.fatalError
         }
```
around line 1880 of Workspace.swift, which meant that callers who wanted to load the package graph would not get anything back if there was any error (i.e. if 1 out of 10 packages had a manifest parse error, the successfully parsed manifests wouldn't be returned, etc).

But when removing that line, then yes, the internal error here failed in the case of a manifest parse error in a root package manifest, since there was no manifest for that package.

Reply:
So the way to trigger this is to have a syntax error in a root manifest, and remove the early exit mentioned above.

Reply:
right, both have been addressed in https://github.com/apple/swift-package-manager/pull/3344/commits/41668932079e2ac4dbfacbd7aba59cf48cdd85d7

Reply:
Great!  Looks like I was reviewing stale diffs.

Reply:
Looks good to me now, curious to hear what @neonichu thinks as well.